File: getting-started-mocks/09-when-handwritten-mocks-stop-scaling/go/notifier/notifier.go
// Package notifier decides whether a customer should hear about their package.
package notifier

import (
	"errors"
	"fmt"
	"time"
)

var (
	ErrCarrierUnavailable = errors.New("carrier unavailable")
	ErrCarrierContract    = errors.New("carrier contract changed")
)

const (
	maxAttempts = 3
	baseBackoff = 100 * time.Millisecond
)

type ShipmentStatus func(trackingNumber string) (string, error)
type Sleep func(time.Duration)
type Sender func(trackingNumber, message string) error

type Recorder interface {
	Record(trackingNumber, message string) error
	Notified(trackingNumber string) bool
}

type Notifier struct {
	status   ShipmentStatus
	send     Sender
	recorder Recorder
	sleep    Sleep
}

func New(status ShipmentStatus, send Sender, recorder Recorder, sleep Sleep) *Notifier {
	return &Notifier{status: status, send: send, recorder: recorder, sleep: sleep}
}

func (n *Notifier) Notify(trackingNumber string) error {
	if n.recorder.Notified(trackingNumber) {
		return nil
	}

	status, err := n.lookup(trackingNumber)
	if err != nil || status != "delayed" {
		return err
	}

	message := fmt.Sprintf("Package %s is delayed", trackingNumber)
	if err := n.send(trackingNumber, message); err != nil {
		return err
	}
	return n.recorder.Record(trackingNumber, message)
}

func (n *Notifier) lookup(trackingNumber string) (string, error) {
	var lastErr error
	for attempt := 1; attempt <= maxAttempts; attempt++ {
		status, err := n.status(trackingNumber)
		if err == nil {
			return status, nil
		}
		if errors.Is(err, ErrCarrierContract) {
			return "", err
		}
		lastErr = err
		if attempt < maxAttempts {
			n.sleep(baseBackoff * time.Duration(1<<(attempt-1)))
		}
	}
	return "", lastErr
}

File: getting-started-mocks/09-when-handwritten-mocks-stop-scaling/node/notifier/notifier.js
import { CarrierContractError } from './errors.js'

export const maxAttempts = 3
const baseBackoffMs = 100

export class Notifier {
  #getShipmentStatus
  #send
  #recorder
  #sleep

  constructor(getShipmentStatus, send, recorder, sleep) {
    this.#getShipmentStatus = getShipmentStatus
    this.#send = send
    this.#recorder = recorder
    this.#sleep = sleep
  }

  async notify(trackingNumber) {
    if (await this.#recorder.notified(trackingNumber)) {
      return
    }
    const status = await this.#lookup(trackingNumber)
    if (status !== 'delayed') {
      return
    }

    const message = `Package ${trackingNumber} is delayed`
    await this.#send(trackingNumber, message)
    await this.#recorder.record(trackingNumber, message)
  }

  async #lookup(trackingNumber) {
    let lastFailure
    for (let attempt = 1; attempt <= maxAttempts; attempt++) {
      try {
        return await this.#getShipmentStatus(trackingNumber)
      } catch (failure) {
        if (failure instanceof CarrierContractError) {
          throw failure
        }
        lastFailure = failure
        if (attempt < maxAttempts) {
          await this.#sleep(baseBackoffMs * 2 ** (attempt - 1))
        }
      }
    }
    throw lastFailure
  }
}
